src/main/java/org/openstreetmap/atlas/checks/validation/tag/TollValidationCheck.java

```diff
@@ -0,0 +1,609 @@
+                if (!this.edgeIntersectsTollFeature(outEdge) && this.containsTollTag(keySet)
+                        && keySet.get(TollTag.KEY).equalsIgnoreCase(TollTag.YES.toString()))
+                {
+                    return this.edgeProvingForwardIsEscapable(outEdge, alreadyCheckedObjectIds);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * @param escapableInEdge
+     *            escapable in edge
+     * @param escapableOutEdge
+     *            escapable out edge
+     * @return boolean for if they are both null
+     */
+    private boolean escapableEdgesNullChecker(final Edge escapableInEdge,
+            final Edge escapableOutEdge)
+    {
+        return escapableInEdge != null && escapableOutEdge != null;
+    }
+
+    /**
+     * @param edge
+     *            an edge
+     * @param alreadyCheckedNearbyTollEdges
+     *            edge that have already been touched when recursing.
+     * @return Id for intersecting toll feature.
+     */
+    private Optional<Long> getAreaOrNodeIntersectionId(final Edge edge,
+            final Set<Long> alreadyCheckedNearbyTollEdges)
+    {
+        alreadyCheckedNearbyTollEdges.add(edge.getIdentifier());
+        final Iterable<Area> intersectingAreas = edge.getAtlas().areasIntersecting(edge.bounds());
+        final Iterable<Node> edgeNodes = edge.connectedNodes();
+        for (final Area area : intersectingAreas)
+        {
+            final boolean areaContainsPolyline = area.asPolygon().overlaps(edge.asPolyLine());
+            final Map<String, String> areaTags = area.getOsmTags();
+            if (areaContainsPolyline && this.containsBarrierTag(areaTags)
+                    && (this.barrierTagContainsToll(areaTags)))
+            {
+                return Optional.of(area.getOsmIdentifier());
+            }
+        }
+
+        for (final Node node : edgeNodes)
+        {
+            final Map<String, String> nodeTags = node.getOsmTags();
+            if ((this.containsHighwayTag(nodeTags) && this.highwayTagContainsToll(nodeTags))
+                    || (this.containsBarrierTag(nodeTags) && this.barrierTagContainsToll(nodeTags)))
+            {
+                return Optional.of(node.getOsmIdentifier());
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * @param edge
+     *            some edge
+     * @return in edges that are car navigable and positive (eliminates reverse edges)
+     */
+    private Set<Edge> getInEdges(final Edge edge)
+    {
+        return edge.inEdges().stream().filter(
+                someEdge -> someEdge.isMainEdge() && HighwayTag.isCarNavigableHighway(someEdge))
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * @param edge
+     *            edge in question
+     * @return nearby toll feature id on the in edge side of the edge in question (upstream)
+     */
+    private Optional<Long> getNearbyTollFeatureInEdgeSide(final Edge edge,
```

Comment:
Is there any chance that this could end up taking a long time to complete, either because the closest toll feature is a long ways away or because this is a depth first search and it takes a while to get to the nearby toll feature.

Reply:
I can just set a limit on how far the search will go. This is just complimentary information for the editor when they are investigating the toll tag modeling. I will limit to around 15 iterations in the search.

Reply:
Same with my other comment about combing two similar methods into one -- I believe this and its outEdge counterpart can be combined